urlc58f2e0b85_xoriole_CEAC_tgz-pJ8-com_sdm_phr_cpabe_CpabeTestJ8/src/com/sdm/phr/bswabe/Bswabe.java

```diff
@@ -0,0 +1,612 @@
+		pub.f.powZn(beta_inv);
+
+		pub.h = pub.g.duplicate();
+		pub.h.powZn(msk.beta);
+
+		pub.g_hat_alpha = pairing.pairing(pub.g, msk.g_alpha);
+	}
+
+	/*
+	 * Generate a private key with the given set of attributes.
+	 */
+	public static BswabePrv keygen(BswabePub pub, BswabeMsk msk, String[] attrs)
+			throws NoSuchAlgorithmException {
+		BswabePrv prv = new BswabePrv();
+		Element g_r, r, beta_inv;
+		Pairing pairing;
+
+		/* initialize */
+		pairing = pub.p;
+		prv.d = pairing.getG2().newElement();
+		g_r = pairing.getG2().newElement();
+		r = pairing.getZr().newElement();
+		beta_inv = pairing.getZr().newElement();
+
+		/* compute */
+		r.setToRandom();
+		g_r = pub.gp.duplicate();
+		g_r.powZn(r);
+
+		prv.d = msk.g_alpha.duplicate();
+		prv.d.mul(g_r);
+		beta_inv = msk.beta.duplicate();
+		beta_inv.invert();
+		prv.d.powZn(beta_inv);
+
+		int i, len = attrs.length;
+		prv.comps = new ArrayList<BswabePrvComp>();
+		for (i = 0; i < len; i++) {
+			BswabePrvComp comp = new BswabePrvComp();
+			Element h_rp;
+			Element rp;
+
+			comp.attr = attrs[i];
+
+			comp.d = pairing.getG2().newElement();
+			comp.dp = pairing.getG1().newElement();
+			h_rp = pairing.getG2().newElement();
+			rp = pairing.getZr().newElement();
+
+			elementFromString(h_rp, comp.attr);
+			rp.setToRandom();
+
+			h_rp.powZn(rp);
+
+			comp.d = g_r.duplicate();
+			comp.d.mul(h_rp);
+			comp.dp = pub.g.duplicate();
+			comp.dp.powZn(rp);
+
+			prv.comps.add(comp);
+		}
+
+		return prv;
+	}
+
+    /*
+     * Delegate a subset of attribute of an existing private key.
+     */
+    public static BswabePrv delegate(BswabePub pub, BswabePrv prv_src, String[] attrs_subset)
+            throws NoSuchAlgorithmException, IllegalArgumentException {
+
+            BswabePrv prv = new BswabePrv();
+            Element g_rt, rt, f_at_rt;
+            Pairing pairing;
+
+            /* initialize */
+            pairing = pub.p;
+            prv.d = pairing.getG2().newElement();
+
+            g_rt = pairing.getG2().newElement();
+            rt = pairing.getZr().newElement();
+            f_at_rt = pairing.getZr().newElement();
+
+            /* compute */
+            rt.setToRandom();
+            f_at_rt = pub.f.duplicate();
+            f_at_rt.powZn(rt);
+            prv.d = prv_src.d.duplicate();
+            prv.d.mul(f_at_rt);
+
+            g_rt = pub.g.duplicate();
+            g_rt.powZn(rt);
+
+            int i, len = attrs_subset.length;
+            prv.comps = new ArrayList<BswabePrvComp>();
+
+            for (i = 0; i < len; i++) {
+                BswabePrvComp comp = new BswabePrvComp();
+                Element h_rtp;
+                Element rtp;
+
+                comp.attr = attrs_subset[i];
+
+                BswabePrvComp comp_src = new BswabePrvComp();
+                boolean comp_src_init = false;
+
+                for (int j = 0; j < prv_src.comps.size(); ++j) {
+                    if (prv_src.comps.get(j).attr == comp.attr) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.